fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolHandlers.java

```diff
@@ -0,0 +1,343 @@
+			if (!(stack.getItem() instanceof DynamicAttributeTool)) {
+				float miningSpeed = stack.getItem().getMiningSpeed(stack, state);
+
+				if (miningSpeed != 1f) {
+					return miningSpeed;
+				}
+			}
+
+			return null;
+		}
+	}
+
+	/**
+	 * This handler handles items that are a subclass of {@link DynamicAttributeTool} by using the
+	 * vanilla {@link Item#isEffectiveOn(BlockState)} with a custom fake tool material to use the mining level
+	 * from {@link DynamicAttributeTool#getMiningLevel(Tag, BlockState, ItemStack, LivingEntity)}.
+	 *
+	 * <p>Only applicable to blocks that are vanilla or share the material that is handled by their vanilla tool.</p>
+	 */
+	private static class ModdedMiningToolsVanillaBlocksToolHandler implements ToolManager.ToolHandler {
+		private final FakeAdaptableToolMaterial fakeMaterial = new FakeAdaptableToolMaterial();
+		private final ImmutableList<ToolItem> vanillaItems;
+
+		private ModdedMiningToolsVanillaBlocksToolHandler(ImmutableList<ToolItem> vanillaItems) {
+			this.vanillaItems = vanillaItems;
+		}
+
+		private ToolItem getVanillaItem(int miningLevel) {
+			if (miningLevel < 0) return vanillaItems.get(0);
+			if (miningLevel >= vanillaItems.size()) return vanillaItems.get(vanillaItems.size() - 1);
+			return vanillaItems.get(miningLevel);
+		}
+
+		@Override
+		public ActionResult isEffectiveOn(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
+			if (stack.getItem() instanceof DynamicAttributeTool) {
+				int miningLevel = ((DynamicAttributeTool) stack.getItem()).getMiningLevel(tag, state, stack, user);
+				if (miningLevel < 0) return ActionResult.PASS;
+				ToolItem vanillaItem = getVanillaItem(miningLevel);
+				ToolMaterial tempMaterial = vanillaItem.getMaterial();
+				fakeMaterial.miningLevel = miningLevel;
+				((ToolItemAccessor) vanillaItem).setMaterial(fakeMaterial);
+				boolean effective = vanillaItem.isEffectiveOn(state) || vanillaItem.getMiningSpeed(new ItemStack(vanillaItem), state) != 1f;
+				((ToolItemAccessor) vanillaItem).setMaterial(tempMaterial);
+				return effective ? ActionResult.SUCCESS : ActionResult.PASS;
+			}
+
+			return ActionResult.PASS;
+		}
+
+		@Override
+		public Float getMiningSpeedMultiplier(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
+			if (stack.getItem() instanceof DynamicAttributeTool) {
+				int miningLevel = ((DynamicAttributeTool) stack.getItem()).getMiningLevel(tag, state, stack, user);
+				if (miningLevel < 0) return null;
+				Item vanillaItem = getVanillaItem(miningLevel);
+				float tempMiningSpeed = ((MiningToolItemAccessor) vanillaItem).getMiningSpeed();
+				((MiningToolItemAccessor) vanillaItem).setMiningSpeed(((DynamicAttributeTool) stack.getItem()).getMiningSpeedMultiplier(tag, state, stack, user));
+				float miningSpeed = vanillaItem.getMiningSpeed(stack, state);
+				((MiningToolItemAccessor) vanillaItem).setMiningSpeed(tempMiningSpeed);
```

Comment:
Is it really needed to go through the vanilla item again? May a pure effectiveness check as above suffice? If not, can't the return value from the unbiased vanilla tool be analyzed and then adjusted for the desired new speed, potentially with some hardcoded values?

Reply:
Also some spacing in that method please, it's stuffed to the brim

Reply:
Don't think we can adjust according to vanilla, for example, swords have constant speed on cobweb no matter the speed of the tool material.

Reply:
These exceptions may be worth hardcoding if it saves us from all the tool manipulation trickery. Maybe first comparing two vanilla tools (like diamond vs iron) to each other is sufficient to test and observe special behavior.

Reply:
In some cases, it should be impossible to compare with two vanilla tools, for example obsidian, it is impossible to have two tools to compare with (in 1.15 anyway), for other levels it should be easy.